packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts

```diff
@@ -35,6 +37,71 @@ function getApisToRemoveForQueryEngine(): Array<GatsbyNodeAPI> {
   return apisToRemove
 }
 
+const getInternalPackagesCacheDir = (): string =>
+  path.join(process.cwd(), `.cache/internal-packages`)
+
+// Create a directory and JS module where we install internally used packages
+const createInternalPackagesCacheDir = async (): Promise<void> => {
+  const cacheDir = getInternalPackagesCacheDir()
+  await fs.ensureDir(cacheDir)
+  await fs.emptyDir(cacheDir)
+
+  const packageJsonPath = path.join(cacheDir, `package.json`)
+
+  await fs.outputJson(packageJsonPath, {
+    name: `gatsby-internal-packages`,
+    description: `This directory contains internal packages installed by Gatsby used to comply with the current platform requirements`,
+    version: `1.0.0`,
+    private: true,
+    author: `Gatsby`,
+    license: `MIT`,
+  })
+}
+
+// Detect if the prebuilt binaries for lmdb have been installed. These are installed under @lmdb and are tied to each platform/arch. We've seen instances where regular installations lack these modules because of a broken lockfile or skipping optional dependencies installs
+function lmdbPrebuiltPackagePresent(): boolean {
+  try {
+    require.resolve(`@lmdb/lmdb-${process.platform}-${process.arch}`)
```

Comment:
In some cases this would not resolve even if it's actually installed when we try to `require(.resolve)` from `gatsby` package context.

Yarn PnP wouldn't allow that because those prebuilt packages are not dependencies of gatsby and are transitive dependency and you can't require/import transitive deps that you package doesn't declare as deps.

In `node_modules` cases it also is possible this won't work if package manager would decide to structure things like this:

```
node_modules
├── gatsby
└── lmdb
    └── @lmdb/lmdb-${process.platform}-${process.arch}
```
this would also mean it doesn't resolve.


For this check to work without relying on `node_modules` hierarchy we need to resolve prebuilt package from `lmdb` package context:

```suggestion
  const lmdbRequire = mod.createRequire(
    require.resolve(`lmdb`)
  )
    lmdbRequire.resolve(`@lmdb/lmdb-${process.platform}-${process.arch}`)
```

The other part is that I'm not sure about is if there are prebuilts for all the platform+arch combinations that exist. If prebuilt for the combo doesn't exist we will always try to install `lmdb` but that wouldn't help, so maybe before we even attempt to install - we should try to check if `@lmdb/lmdb-${process.platform}-${process.arch}` is included in list of available prebuilts (gathered from lmdb's optional dependencies list):
```
[
  '@lmdb/lmdb-darwin-arm64', 
  '@lmdb/lmdb-darwin-x64', 
  '@lmdb/lmdb-linux-arm', 
  '@lmdb/lmdb-linux-arm64', 
  '@lmdb/lmdb-linux-x64', 
  '@lmdb/lmdb-win32-x64'
]
```

Reply:
Greate suggestions, I'll add those in 👍